### DISCUSSION TITLE
How to deprecate content

### DISCUSSION THREAD

**Discussion by Ryuno-Ki** [mdn/mdn-community#199] [Content]
This will resolve in https://github.com/mdn/yari/issues/2903 as actionable summary.

From what I understand, at the moment, deprecation is done via macros, that render some EJS template.

Since we want to migrate to Markdown as main format for editing in the future, it's worth thinking about how we would deal with it in that world.
In my mind, that could be handled by a line in Frontmatter (`deprecated: true` or `deprecated: 2021-02-14` or `deprecated: ['Firefox 85', 'Chrome 90']`) and then handled by Yari when generating the markup.

How would you approach it?

**Comment by ghost** [comment]
Sounds interesting to me.
In my opinion, we must distinguish from what is deprecated by the W3C and what is no longer supported by a browser. We don't risk creating too complicated data models if we know how the business logic works, us or others.
And above all even if we don't use Typescript on YALM we still try to be consistent with the data types spontaneously.
For example, in this case I would associate the word `deprecated` with the equivalent of a `true` or `false` in YALM, no Array or date. Rather for other data we add keys.

**Comment by peterbe** [comment]
> From what I understand, at the moment, deprecation is done via macros, that render some EJS template.

What macro is that?

**Reply by Ryuno-Ki** [reply]
https://github.com/mdn/yari/blob/9e3e569ce8df374a50e343da791eb245420106e4/kumascript/macros/Deprecated_Inline.ejs

as mentioned by @hamishwillee in https://github.com/mdn/content/issues/2086#issuecomment-777892990

**Reply by ghost** [reply]
Pretending that we take everything from [these JSON](https://www.github.com/mdn/browser-compat-data/tree/master/html%2Felements%2Fbase.json) (browsers-compact-data) as an example, the macro is no longer needed and should be deleted?

**Comment by ghost** [comment]
https://www.w3.org/TR/html401/conform.html#:~:text=A%20deprecated%20element%20or%20attribute,in%20future%20versions%20of%20HTML.

According to this document, the `deprecaded` property should belong to the `Element` and `Attribute` class objects intended as articles in MDN that talk about an element or an attribute in HTML. 
The note could simply state the status and reference the definition on the W3C specification unless there is an MDN article of its own that talks about this status.

Examples
- The feature documented in this article is [deprecated](https://www.w3.org/TR/html401/conform.html#:~:text=A%20deprecated%20element%20or%20attribute,in%20future%20versions%20of%20HT).
- The element documented in this article is [deprecated](https://www.w3.org/TR/html401/conform.html#:~:text=A%20deprecated%20element%20or%20attribute,in%20future%20versions%20of%20HT).
- The attribute documented in this article is [deprecated](https://www.w3.org/TR/html401/conform.html#:~:text=A%20deprecated%20element%20or%20attribute,in%20future%20versions%20of%20HT).

---

It is perhaps not necessary to create a class for each type of `Article` but perhaps also just an article class that has an enumerable property of type (`attribute`, `element`, `property`, `method`, `function`, `value`). We need to think, we need the _people of content_. An actionable issue could be supporting the key in YALM or somewhere right to define the properties of the article object (a JSON, a meta tag if that's valid). Here too to reflect.

---

Also I was wondering if we can take the data from [BCD](https://www.github.com/mdn/browser-compat-data/tree/master/html%2Felements%2Fbase.json) in order to be consistent, we should analyze if they have it and redo a logic similar to the BCD table.

---

```JavaScript
const MDN = { 
class Article {
  static title = new String();
  static content = new MDN.Content(); // 🤏
  static slug = new MDN.Slug(); // 👎
},
class Element extend MDN.Article {
  static deprecated = false;
 }
}
const article = new MDN.Element();
article.deprecated = YALM.deprecated;
```
Pseudo code ⬆️ (attempt to [namespaces](https://www.google.com/amp/s/www.geeksforgeeks.org/javascript-namespace/amp/) because the classes are pretty the same as the real Javascript classes documented 😑)

---

But at the moment there is a component called `Article` at the beginning of the template which you can add this `if`?

**Reply by Ryuno-Ki** [reply]
Nitpick: It's YAML - for Yet Another Markup Language.

**Reply by ghost** [reply]
Yep.
Sorry.
One of the reasons it doesn't convince me is because it doesn't have a page or class on MDN.
Instead JSON yes.
This to me layman means potentially less hassle and less string manipulation.

- https://developer.mozilla.org/en-US/search?q=YAML
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON

**Reply by Ryuno-Ki** [reply]
Nothing is stopping you from adding it to a Glossary page or so :-)

**Reply by ghost** [reply]
In any case, do we already have a JavaScript representation of the `Frontmatter`?

**Reply by ghost** [reply]
> Nothing is stopping you from adding it to a Glossary page or so :-)

For the moment JSON, XML and SQL are enough for me but I might change my mind. I see that GitHub also uses it for actions. I wait for an _engineering intervention from above_ to understand if it is really recommended (Especially examples for parising in JavaScript.). 

---

However, returning to the subject, does the decision here in this discussion also affect the CLI?

**Reply by ghost** [reply]
FTR
https://en.m.wikipedia.org/wiki/Single_source_of_truth

**Comment by hamishwillee** [comment]
I think it is better to generate lists of things like this dynamically (as stated elsewhere).
- The dynamic list doesn't necessarily have to work for every case to be better than what we have now.
- I'm not sure that this can be done as list from tags/metadata due to granularity - but perhaps it can (after all, even attributes often have their on page on MDN). [FWIW, I'd like dynamic lists from tags anyway :-) )
- BBD as suggested by Mattia - great idea, but BCD definitely doesn't go to the level of detail needed (even aside from other technical issues of integrating two markups.
- I'm OK for macros for this kind of thing - you have formalised markup you want somewhere anyway, and you slurp that up for lists. I don't mind that it doesn't render perfectly in markdown. 

I'm hoping the dev team can give us some kind of direction here on what is reasonable to achieve with the tooling.

PS. Regarding "consensus" @peterbe - I presume you mean with the idea that dynamically generating a list like this is "a good idea"? Or do you mean with the precise approach?

**Reply by ghost** [reply]
Are You talking about validating HTML content against a kind of [schema](https://schema.org) so we can have structured data directly from there?
I know that Is common in [XML](https://en.m.wikipedia.org/wiki/XML_validation).
I can't imagine the whole thing well but the fact that a data is entered in only 1 point is good, even if this point is an HTML table formatted in a certain way or whatever.

**Comment by wbamberg** [comment]
FWIW I think we should use BCD for this. We already represent "deprecated" in BCD, representing it in two different places is a recipe for inconsistency. 

One possibility here (in some structured content paradise) is that pages include the BCD query (e.g. "api.Navigator.sendBeacon") in the front matter. When the page is built, we resolve that reference and use it to build the BCD table and to add whatever markup we want to use to indicate "deprecated" (and anything else we could derive from BCD, like spec URLs).

> BCD definitely doesn't go to the level of detail needed

I'm not sure I understand this, can you give an example?

**Reply by wbamberg** [reply]
With https://github.com/mdn/yari/issues/3258 landing, we're close to being able to do this!

**Reply by Ryuno-Ki** [reply]
The PR landed. Is there a way to deprecate content now?
(I'm still catching up with the last two months, but noticed some discussions around Markdown editing)

**Reply by peterbe** [reply]
We have the mdn/archived-content trick in Yari. But the current plan is to stop that entirely and stop hosting pages built from it. 

So the right thing to do if you want "deprecate" a piece of content is to simply delete it. But perhaps it might be a smart thing to first dump a copy of it in mdn/archived-content instead of relying on forensically going through git deletion logs in mdn/content (and mdn/translated-content)

Perhaps the next action is to write a CLI too to Yari that helps you with all of this stuff. It would assist you with rendering it out one last time and the automate the copying over to your `$CONTENT_ARCHIVED_ROOT` followed by executing a `git rm -r ...` and perhaps doing whatever's needed to the `_redirects.txt`.

**Reply by Ryuno-Ki** [reply]
That would be great!

The last thing it should output is a reminder to commit and uploaded the content changed in the archived repo clone created this way.

**Reply by wbamberg** [reply]
I'm confused now. I thought this thread was about marking Web platform features as deprecated in the docs, not about deprecating/archiving docs pages themselves.

**Reply by peterbe** [reply]
> I'm confused now. I thought this thread was about marking Web platform features as deprecated in the docs, not about deprecating/archiving docs pages themselves.

Oh sorry. I might have been confused on the topic. Sorry.

**Reply by hamishwillee** [reply]
@peterbe Yes, so this one is about how we mark features as deprecated. Right now we do it in BCD, in the affected feature doc (via `{{deprecated_header}}`), and in the parent of the feature (via `{{deprecated_inline}}` or whatever it is). Now that we have the ability to specify the BCD as metadata, the hope @wbamberg is expressing is that we might be able to automatically insert the deprecation markup in the generated content from BCD definitions.

**Reply by wbamberg** [reply]
To start with it's worth distinguishing what we've sometimes called:

* "banners", that are block-level things that appear along the top of the page they pertain to, like https://developer.mozilla.org/en-US/docs/Web/API/Screen/onorientationchange. Implemented using something like [`deprecated_header`](https://github.com/mdn/yari/blob/main/kumascript/macros/Deprecated_Header.ejs).
* "badges", that are inline things, usually icons, that appear next to links to the thing they pertain to. For, example in the `applets` link in the sidebar here: https://developer.mozilla.org/en-US/docs/Web/API/Document. Implemented using something like [`deprecated_inline`](https://github.com/mdn/yari/blob/main/kumascript/macros/Deprecated_Inline.ejs).

## Banners

For the "deprecated" banner that appears in pages like https://developer.mozilla.org/en-US/docs/Web/API/Screen/onorientationchange, we could do something like this:

* Simultaneously:
    * implement something in Yari that, if there's a browser-compat front matter item, and it has `deprecated` set, adds the `deprecated` banner to the page
    * update `deprecated_header` to be a no-op if there's a browser-compat front matter item
* Then at our leisure:
    * go through all our docs, moving the browser-compat query from the Compat macro into the front matter
* Finally, retire the `deprecated_header` macro.

We could also do the same for the "Experimental" banners that get added by the `SeeCompatTable` macro in pages like https://developer.mozilla.org/en-US/docs/Web/API/PaymentRequest/shippingOption.

There's more discussion of this general idea in https://github.com/mdn/stumptown-content/issues/306, from a slightly different world though.

## Badges

These are a bit different. In cases where the content is built (like sidebars), we could imagine doing something similar. If the sidebar builder is given the path to the file representing the page to link, it can also look in the front matter/bcd to get this value. Would be horrible for build performance though :). Or we could just not bother adding these icons to sidebars.

There are also cases where these macros are added by handwritten prose. These are usually lists of properties/methods, like https://developer.mozilla.org/en-US/docs/Web/API/BluetoothDevice#properties. Ideally we might eventually be able to generate these lists of links too, if we had a way to represent, say, the methods of an API, and then we could do the same thing there. But that's for the future.

**Reply by peterbe** [reply]
@wbamberg Regarding that rough plan for banners; would it be best to keep the `{{deprecated_header}}` string inside the doc? But what we could do, inside the `deprecated_header.ejs` code, is have access to the resolved BCD data and use that to determine whether to display or just return an empty string. Then, you can at least have full control of the placement of the banner. 

I guess the other option is that Yari would take a stance and automatically inject the banner based on some business logic of parsing the `browser-compat` from the front-matter. But I guess its only option would be to inject it **before any of the prose sections** from the parsed KS-HTML (or Markdown).

**Reply by peterbe** [reply]
> it can also look in the front matter/bcd to get this value. Would be horrible for build performance though :).

It might actually not be too bad. That stuff would be cached and stuff. We're most likely going to need to render those sub-pages anyway so with caching it wouldn't really matter if you do it when you render the page, or you render a page whose sidebar trigger that lookup. 
I wouldn't count it out as too impossible.

**Reply by wbamberg** [reply]
> @wbamberg Regarding that rough plan for banners; would it be best to keep the `{{deprecated_header}}` string inside the doc? But what we could do, inside the `deprecated_header.ejs` code, is have access to the resolved BCD data and use that to determine whether to display or just return an empty string. Then, you can at least have full control of the placement of the banner.
> 
> I guess the other option is that Yari would take a stance and automatically inject the banner based on some business logic of parsing the `browser-compat` from the front-matter. But I guess its only option would be to inject it **before any of the prose sections** from the parsed KS-HTML (or Markdown).

I think the banner should always be at the top, so it is better for Yari to inject it automatically. That's actually one of the big advantages of this proposal IMO - authors don't have to remember to include this macro and to put it in the right place. In Kuma, everything was done ad-hoc in content - banners, sidebars, next/previous links. In Yari it would be great to get away from this. I think of these banners (and sidebars too) as being more like breadcrumbs, say, than like page content.

**Reply by peterbe** [reply]
Excellent. I'm in favor of that. 
Let me know via mdn/yari issues when I can help.

**Comment by hamishwillee** [comment]
>> BCD definitely doesn't go to the level of detail needed

> I'm not sure I understand this, can you give an example?

I'm sure I had a specific case when I wrote that but I can't think of what it was now. There are at least some APIs though where the feature docs don't cover all functions, so there is nowhere  to mark that a particular function is deprecated.

HTTP methods are commonly not in BCD if there is no clear meaning for "supported by a browser", so there would be nowhere to mark those if deprecated.

There are probably other things BCD does not cover at all but we still need to mark deprecation in docs.

Which is not to say I don't completely agree with at least attempting to do this. It's just that it might mean we have to document to greater breadth and depth in BCD than we have previously chosen to do.

**Comment by pepelsbey** [comment]
This discussion comes from the deep past, so it has most likely lost its relevance due to changes in content/yari since then.

Feel free to start a new one if you still want to address the topic.
